README.md
# LABOR - Asset Building - Dev Server Extension
**This package is an extension to our [asset building bundle](https://www.npmjs.com/package/@labor-digital/asset-building).**

When it comes to developing Single page applications (SPAs), you will find out sooner or later that it is convenient to have a local dev server that supports stuff like "hot-module-replacement".

The problem with webpack's default dev-server was, it does not work correctly when multiple apps run in the same domain (frontend and backend, for example). It also only works with real "webpack.js" files and not with the dynamic version we use.
To circumvent that, and to make the integration as smooth as possible, we created our own dev server using express and webpack's middleware capabilities.

## Installation
* Install our asset builder:
``` 
npm install --save-dev @labor-digital/asset-building
```
* Install the npm dependency:
```
npm install --save-dev @labor-digital/asset-building-dev-server
```
* Add the provider plugin to your package.json
```
{ 
    "builderVersion": 2,
    [...]
    "extensions": [
        "@labor-digital/asset-building-dev-server"
    ]
}
```
* Add the following part to your package.json:
```
  "scripts": {
    [...]
    "dev": "labor-asset-building dev"
  }
```
* Start your server by running "npm run dev" -> The urls are printed to the console
* Done! :-)

## Adding the dev server to express apps
Our asset builder can run as a middleware in express, providing
hot module replacement and webpack's dev features in the same way the "dev" command does for virtually
every express app.

**We provide a function you can use in your server.js file. It will only apply
the required middlewares if your NODE_ENV variable is set to "development".**

Register the express plugin in your index.js like:
```javascript
const expressAssetBuildingPlugin = require("@labor-digital/asset-building/dist/Interop/Express/expressAssetBuildingPlugin.js");
const expressDevServerPlugin = require("@labor-digital/asset-building-dev-server/dist/ExpressDevServerPlugin");
const express = require("express");
const app = express();

// Apply the dev server plugin to the app
expressAssetBuildingPlugin(app)
   .then(expressDevServerPlugin)
    .then(() => {
        // Your express - app setup goes here...
    });
```

## Changing the visible URL in the console
By default, the console will show a URL like http://js.localmachine.space:9999 when you are launching a new instance of a project. If you want to change the URL, you can do so by adding the "devUrl" option to your labor configuration in your package.json. The {{port}} placeholder
will automatically be replaced with the correct port of your application.
```
{
    [...]
    "labor": {
        "devUrl": "http://yourUrl:{{port}}"
    }
}
```

## Whats xyz.localmachine.space?
To work locally, you can work with an IP like 127.0.0.1, localhost, or you can point a DNS record to your local machine. 
The localmachine.space domain is precisely that. It is a domain name that always leads to "127.0.0.1", no matter what subdomain you use.
Its a helper to make dev URLs more comfortable to read; it works on every machine as long as you have an internet connection.

## Postcardware
You're free to use this package, but if it makes it to your production environment, we highly appreciate you sending us a postcard from your hometown, mentioning which of our package(s) you are using.

Our address is: LABOR.digital - Fischtorplatz 21 - 55116 Mainz, Germany.

We publish all received postcards on our [company website](https://labor.digital). 


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file. See [standard-version](https://github.com/conventional-changelog/standard-version) for commit guidelines.

## [2.0.0](https://github.com/labor-digital/asset-building-dev-server/compare/v1.8.0...v2.0.0) (2020-10-23)


### ⚠ BREAKING CHANGES

* major dependency update

### Features

* adjust plugin for new asset builder api ([cbfa513](https://github.com/labor-digital/asset-building-dev-server/commit/cbfa513863917a5d433f391a0dd1eb41db757307))
* update to asset-builder v8 ([2b6aa26](https://github.com/labor-digital/asset-building-dev-server/commit/2b6aa2623dd112f700aa0ca56919dc98cb00d498))

## [1.8.0](https://github.com/labor-digital/asset-building-dev-server/compare/v1.7.1...v1.8.0) (2020-07-10)


### Features

* make code PSR-2 compliant ([25820f7](https://github.com/labor-digital/asset-building-dev-server/commit/25820f71e7e4c79bc6d4726385db221000ebaffb))
* update dependencies and package.json + fix express docs ([56f5a99](https://github.com/labor-digital/asset-building-dev-server/commit/56f5a99c5f65945ba22dce8fc75f999a98289bdf))


### Bug Fixes

* update dependencies ([603842e](https://github.com/labor-digital/asset-building-dev-server/commit/603842e0db21e71e984ec5787b4961e95751d4bf))
* use correct ExpressContext ([e0f9eda](https://github.com/labor-digital/asset-building-dev-server/commit/e0f9eda9839da381214986ea8d40b833fe69fe6e))

### [1.7.1](https://github.com/labor-digital/asset-building-dev-server/compare/v1.7.0...v1.7.1) (2020-04-13)


### Bug Fixes

* update dependencies + fix typos in the readme.md ([21b1e41](https://github.com/labor-digital/asset-building-dev-server/commit/21b1e415c030dc73ff33575289925a24f999b8be))

## [1.7.0](https://github.com/labor-digital/asset-building-dev-server/compare/v1.6.2...v1.7.0) (2020-04-11)


### Features

* update dependencies ([cae907b](https://github.com/labor-digital/asset-building-dev-server/commit/cae907b0419d5238d10a1451b00c748b0bedf634))

### [1.6.2](https://github.com/labor-digital/asset-building-dev-server/compare/v1.6.1...v1.6.2) (2020-03-23)

### [1.6.1](https://github.com/labor-digital/asset-building-dev-server/compare/v1.6.0...v1.6.1) (2020-02-17)

## [1.6.0](https://github.com/labor-digital/asset-building-dev-server/compare/v1.5.1...v1.6.0) (2020-02-17)


### Features

* preparation for github release ([ef10b18](https://github.com/labor-digital/asset-building-dev-server/commit/ef10b18a2561e9c56f0821a7ac035a5d12498738))
* update dependencies and add cache-control headers to the dev middleware ([2f04ced](https://github.com/labor-digital/asset-building-dev-server/commit/2f04cedf252fb049c92856b630d98c06eadc5d6e))


### Bug Fixes

* update [@labor](https://github.com/labor) reference in link command ([13b3f2a](https://github.com/labor-digital/asset-building-dev-server/commit/13b3f2afa82f22b2cb9cd216363e981c73e08a59))

## [1.5.1](https://bitbucket.org/labor-digital/labor-dev-assetbuilding-dev-server/branches/compare/v1.5.1%0Dv1.5.0#diff) (2019-11-08)


### Bug Fixes

* make sure the dev server keeps running even if webpack failed to build the bundle ([74c87e4](https://bitbucket.org/labor-digital/labor-dev-assetbuilding-dev-server/commits/74c87e4))



# [1.5.0](https://bitbucket.org/labor-digital/labor-dev-assetbuilding-dev-server/branches/compare/v1.5.0%0Dv1.4.0#diff) (2019-10-15)


### Features

* update to latest asset-building package ([a858eaa](https://bitbucket.org/labor-digital/labor-dev-assetbuilding-dev-server/commits/a858eaa))



# [1.4.0](https://bitbucket.org/labor-digital/labor-dev-assetbuilding-dev-server/branches/compare/v1.4.0%0Dv1.3.0#diff) (2019-10-14)


### Features

* add express dev server plugin ([ef7a890](https://bitbucket.org/labor-digital/labor-dev-assetbuilding-dev-server/commits/ef7a890))



# [1.3.0](https://bitbucket.org/labor-digital/labor-dev-assetbuilding-dev-server/branches/compare/v1.3.0%0Dv1.2.1#diff) (2019-10-14)


### Features

* update to latest asset builder version ([b7d3279](https://bitbucket.org/labor-digital/labor-dev-assetbuilding-dev-server/commits/b7d3279))



## [1.2.1](https://bitbucket.org/labor-digital/labor-dev-assetbuilding-dev-server/branches/compare/v1.2.1%0Dv1.2.0#diff) (2019-09-26)


### Bug Fixes

* redirect history api to the correct entry point ([b508d0c](https://bitbucket.org/labor-digital/labor-dev-assetbuilding-dev-server/commits/b508d0c))



# [1.2.0](https://bitbucket.org/labor-digital/labor-dev-assetbuilding-dev-server/branches/compare/v1.2.0%0Dv1.1.0#diff) (2019-09-26)


### Features

* implement history api fallback middleware ([612e480](https://bitbucket.org/labor-digital/labor-dev-assetbuilding-dev-server/commits/612e480))



# [1.1.0](https://bitbucket.org/labor-digital/labor-dev-assetbuilding-dev-server/branches/compare/v1.1.0%0Dv1.0.1#diff) (2019-06-13)


### Features

* update dependencies and add pipeline support ([990019d](https://bitbucket.org/labor-digital/labor-dev-assetbuilding-dev-server/commits/990019d))



## [1.0.1] - 2019-01-10
### Added
- Added eventsource-polyfill to support hot module reload in IE and Safari

## [1.0.0] - 2018-12-18
Initial commit

The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/)
and this project adheres to [Semantic Versioning](http://semver.org/spec/v2.0.0.html).


flushLib.js
/*
 * Copyright 2019 LABOR.digital
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 * Last modified: 2019.03.18 at 15:12
 */
const fs = require('fs');
const path = require('path');

function rmdirRecursive(dirname, removeSelf)
{
    if (fs.existsSync(dirname)) {
        fs.readdirSync(dirname).forEach(function (file, index) {
            var curPath = dirname + path.sep + file;
            if (fs.lstatSync(curPath).isDirectory()) { // recurse
                rmdirRecursive(curPath);
            } else { // delete file
                fs.unlinkSync(curPath);
            }
        });
        if (removeSelf !== false) {
            fs.rmdirSync(dirname);
        }
    }
}

try {
    rmdirRecursive(__dirname + '/dist', false);
} catch (e) {
    console.error('Failed to clear dist!');
}

plugins/DevServerPlugin.js
/**
 * Created by Martin Neundorfer on 18.12.2018.
 * For LABOR.digital
 */
module.exports = class DevServerPlugin {

	getModes(modes) {
		let hasDev = false;
		modes.forEach(mode => {
			if (mode === "dev") hasDev = true;
		});
		if (!hasDev) modes.push("dev");
		return modes;
	}

	isProd(isProd, mode) {
		return isProd ? mode !== "dev" : isProd;
	}

	filterLaborConfig(laborConfig, context) {
		if (context.mode !== "dev") return laborConfig;
		if (laborConfig.builderVersion < 2) throw new Error("The dev server only works with build version 2 apps!");
		return laborConfig;
	}

	isComponentEnabled(enabled, component, context) {
		if (context.mode !== "dev") return enabled;

		// Enable HtmlPlugin
		if (component === "HtmlPlugin") return true;
		// Disable output cleaner plugin
		if (component === "CleanOutputDirPlugin") return false;
		return enabled;
	}

	/**
	 *
	 * @param {module.ConfigBuilderContext} context
	 * @param key
	 * @return {*}
	 */
	afterComponent(context, key) {
		if (context.mode !== "dev") return context;

		// Make sure at least the default template is enabled
		if (typeof context.currentAppConfig.htmlTemplate === "undefined")
			context.currentAppConfig.htmlTemplate = true;

		return context;
	}

	alternativeCompiler(useDefaultCompiler, webpack, callback, context) {
		if (context.mode !== "dev") return useDefaultCompiler;

		// Prepare the server
		const http = require("http");
		const express = require("express");
		const app = express();
		const MiscHelpers = require("@labor/asset-building/src/Helpers/MiscHelpers");
		const portfinder = require("portfinder");

		// Get a new port for the server to work on
		portfinder.getPortPromise({
			port: 8080,
			stopPort: 8150
		}).then(port => {
			// Keep the callback alive
			context.laborConfig.keepAlive = true;

			// Render a header
			console.log("");
			console.log("Dev-Server started and is now listening on " + port + "!");
			console.log(MiscHelpers.drawLine());
			context.webpackConfig.forEach(config => {
				console.log("* APP - " + config.name + ": http://js.localmachine.space:" + port + "/app-" + config.name + "/");
			});
			console.log(MiscHelpers.drawLine());

			// Creating webpack instances
			context.webpackConfig.forEach(config => {
				// Rewrite element config
				config = this._rewriteAppConfig(config, context);

				// Create wrapper for stat rendering
				// This simulates the "child" rendering that is used in our asset-building strategy
				const statWrapper = function (stats) {
					return {
						toJson: function () {
							const c = stats.toJson(...arguments);
							c.name = config.name;
							return {
								children: [c]
							}
						}
					}
				};

				// Create a compiler wrapper to use our own callbacks
				const compiler = webpack(config);
				compiler.__run = compiler.run;
				compiler.run = function (c, callback) {
					return compiler.__run(c, function (e, s) {
						context.callback(context, e, statWrapper(s));
						return callback.call(compiler, ...arguments);
					});
				};
				compiler.__watch = compiler.watch;
				compiler.watch = function (c, callback) {
					return compiler.__watch(c, function (e, s) {
						context.callback(context, e, statWrapper(s));
						return callback.call(compiler, ...arguments);
					});
				};

				// Register fallback for history router
				app.use(require("connect-history-api-fallback")({
					"index": "/app-" + config.name
				}));

				// Register webpack as express middleware
				app.use(require("webpack-dev-middleware")(compiler, {
					logLevel: "silent",
					publicPath: config.output.publicPath
				})).use(require("webpack-hot-middleware")(compiler, {
					// log: false,
					path: config.output.publicPath + "__webpack_hmr",
					heartbeat: 10 * 1000
				}));

			});

			// Start the server
			const server = http.createServer(app);
			server.listen(port);
		}).catch(err => {
			throw new Error(err);
		});


		// Done -> True disables default compiling
		return false;
	}

	/**
	 * This helper is used to receive the webpack config of a single app which is then
	 * restructured to work with the hot module replacement middleware
	 *
	 * @param config
	 * @param {module.ConfigBuilderContext} context
	 * @return {*}
	 * @private
	 */
	_rewriteAppConfig(config, context) {
		// Rewrite app configuration
		const appId = "app-" + config.name;
		const appConfig = context.laborConfig.apps[parseInt(config.name)];

		delete config.output.path;
		config.output.publicPath = "/" + appId;

		// Rewrite entry to inject additional scripts
		if(typeof config.entry === "string") config.entry = [config.entry];
		if(Array.isArray(config.entry)) config.entry = {main: config.entry};
		if(typeof config.entry !== "object" || !Array.isArray(config.entry.main)) throw new Error("Invalid entry configuration!");
		config.entry.main.unshift("webpack-hot-middleware/client?path=/" + appId + "__webpack_hmr&timeout=20000&reload=true");
		config.entry.main.unshift("eventsource-polyfill");

		// Disable git adding
		appConfig.disableGitAdd = true;

		// Require plugins only if required
		const webpack = require("webpack");

		// Require support for hot module replacement
		config.plugins.push(new webpack.HotModuleReplacementPlugin());

		// Done
		return config;
	}
};

src/expressDevServerPlugin.ts
import ExpressContext from '@labor-digital/asset-building/dist/Interop/Express/ExpressContext';

declare global
{
    namespace NodeJS
    {
        interface Global
        {
            EXPRESS_DEV_SERVER_PLUGIN_MODE: boolean
        }
    }
}

// Mark plugin as loaded
global.EXPRESS_DEV_SERVER_PLUGIN_MODE = true;

module.exports = function expressDevServerPlugin(context: ExpressContext): Promise<ExpressContext> {
    
    // Ignore if this is not a dev context
    if (context.isProd) {
        console.log('expressDevServerPlugin: Skipping dev server, because the app is running in production mode!');
        return Promise.resolve(context);
    }
    
    // Register webpack as express middleware
    context.expressApp.use(require('webpack-dev-middleware')(context.compiler, {
        logLevel: 'silent',
        publicPath: context.compiler.options.output.publicPath.replace(/^\./, ''),
        headers: {
            etag: null,
            'cache-control': 'no-cache,no-store',
            'Pragma': 'no-cache'
        }
    }));
    context.expressApp.use(require('webpack-hot-middleware')(context.compiler, {
        // log: false,
        path: '/__webpack_hmr',
        heartbeat: 10 * 1000
    }));
    
    // Done
    return Promise.resolve(context);
};

src/DevServer.ts
import {AssetBuilderEventList} from '@labor-digital/asset-building/dist/AssetBuilderEventList';
import {CoreContext} from '@labor-digital/asset-building/dist/Core/CoreContext';
import {WorkerContext} from '@labor-digital/asset-building/dist/Core/WorkerContext';

/**
 * Helper which draws a line for "design" purposes
 * @param char
 */
function drawLine(char?: string): string
{
    char = typeof char === 'string' ? char : '=';
    return char.repeat(90);
}

/**
 * This helper is used to rewrite the webpack app configuration to match the requirements
 * for the hot module replacement middleware
 */
function rewriteAppConfig(context: WorkerContext): void
{
    const config = context.webpackConfig;
    config.watch = true;
    
    // Rewrite entry to inject additional scripts
    if (typeof config.entry === 'string') {
        config.entry = [config.entry];
    }
    if (Array.isArray(config.entry)) {
        config.entry = {main: config.entry};
    }
    if (typeof config.entry !== 'object' || !Array.isArray(config.entry.main)) {
        throw new Error(
            'Invalid entry configuration!');
    }
    config.entry.main.unshift('webpack-hot-middleware/client?path=/__webpack_hmr&timeout=20000&reload=true');
    config.entry.main.unshift('eventsource-polyfill');
    
    // Disable git adding
    context.app.disableGitAdd = true;
    
    // Require support for hot module replacement
    config.plugins.push(new (require('webpack')).HotModuleReplacementPlugin());
}

export default function (context: CoreContext, scope: string) {
    if (scope !== 'global') {
        throw new Error('The dev server extension has to be registered on a global scale!');
    }
    if (context.builderVersion !== 2) {
        throw new Error('The dev server only works with builder version 2 definitions!');
    }
    
    // Add the dev mode into the config
    context.eventEmitter.bind(AssetBuilderEventList.GET_MODES, (e) => {
        e.args.modes.push('dev');
    });
    
    // Mark the process as non production when the dev server is active
    context.eventEmitter.bind(AssetBuilderEventList.IS_PROD, (e) => {
        e.args.isProd = e.args.mode === 'dev' ? false : e.args.isProd;
    });
    
    // Set the mode to dev if we are running in express context
    context.eventEmitter.bind(AssetBuilderEventList.GET_MODE, (e) => {
        if (global.EXPRESS_DEV_SERVER_PLUGIN_MODE === true &&
            process.env.NODE_ENV === 'development') {
            e.args.mode = 'dev';
        }
    });
    
    // Register our custom compiler
    context.eventEmitter.bind(AssetBuilderEventList.FILTER_WEBPACK_COMPILER, (e) => {
        // Ignore if we are not in dev mode
        const context: WorkerContext = e.args.context;
        if (context.mode !== 'dev') {
            return;
        }
        
        // Check if we are running as plugin in an existing app
        if (context.parentContext.environment === 'express') {
            // Rewrite element config
            rewriteAppConfig(context);
            return;
        }
        
        // Check if the app disables the dev server
        if (context.app.disableDevServer) {
            return;
        }
        
        // Stop the propagation
        e.stopPropagation();
        const webpack = e.args.compiler;
        
        // Add a foo compiler
        e.args.compiler = function () {
            return true;
        };
        
        // Return a promise while we are starting our server
        return new Promise((resolve, reject) => {
            // Prepare the server
            const http = require('http');
            const express = require('express');
            const expressApp = express();
            const portFinder = require('portfinder');
            
            // Get a new port for the server to work on
            portFinder.getPortPromise({
                port: 8080 + context.appId * 10,
                stopPort: 8150
            }).then(port => {
                
                // Prepare the url
                let url = 'http://js.localmachine.space:{{port}}';
                if (typeof (context.parentContext.laborConfig as any).devUrl === 'string') {
                    url = (context.parentContext.laborConfig as any).devUrl;
                }
                url = url.replace(/{{port}}/g, port);
                
                // Render a header
                console.log('');
                console.log('Dev-Server started and is now listening on ' + port + '!');
                console.log(drawLine());
                console.log('* APP - ' + context.app.appName + ': ' + url + '/');
                console.log(drawLine());
                
                // Rewrite element config
                rewriteAppConfig(context);
                
                // Create the webpack instance
                const compiler = webpack(context.webpackConfig, (err, stats) => {
                    // Check if we got obvious errors
                    if (err !== null) {
                        return reject(err);
                    }
                    e.args.callback(context, stats, () => null, reject);
                });
                
                // Register fallback for history router
                expressApp.use(require('connect-history-api-fallback')({
                    'index': '/'
                }));
                
                // Register webpack as express middleware
                expressApp.use(require('webpack-dev-middleware')(compiler.compiler, {
                    logLevel: 'silent',
                    publicPath: '/'
                })).use(require('webpack-hot-middleware')(compiler.compiler, {
                    // log: false,
                    path: '/__webpack_hmr',
                    heartbeat: 10 * 1000
                }));
                
                // Start the server
                const server = http.createServer(expressApp);
                server.listen(port);
                resolve();
            }).catch(e.args.reject);
        });
    });
}